src/ast_editor.ml

```diff
@@ -0,0 +1,578 @@
+    | ".ml" -> Structure
+    | ".mli" -> Signature
+    | _ -> Unknown
+
+  let get_pp_path ~(document : TextDocument.t) =
+    let relative = relative_document_path ~document in
+    match project_root_path () with
+    | None ->
+      let (_ : _ Promise.t) =
+        Window.showErrorMessage ~message:"Project root wasn't found." ()
+      in
+      None
+    | Some root ->
+      let build_root = "_build/default" in
+      let fname =
+        match get_kind ~document with
+        | Unknown -> failwith "Unknown file extension"
+        | Structure -> String.chop_suffix_exn ~suffix:".ml" relative ^ ".pp.ml"
+        | Signature ->
+          String.chop_suffix_exn ~suffix:".mli" relative ^ ".pp.mli"
+      in
+      let ( / ) = Caml.Filename.concat in
+      Some (root / build_root / fname)
+end
+
+let fetch_pp_code ~document =
+  match Pp_path.get_pp_path ~document with
+  | None ->
+    show_message `Error "Unable to find preprocessed file for %s"
+      (Uri.toString (TextDocument.uri document) ());
+    ""
+  | Some path -> (
+    match Ppx_tools.get_reparsed_code_from_pp_file ~path with
+    | Ok code -> code
+    | Error err_msg ->
+      show_message `Error "%s" err_msg;
+      "")
+
+let transform_to_ast instance ~(document : TextDocument.t)
+    ~(webview : WebView.t) =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let open Ppx_tools in
+  let make_value result =
+    match result with
+    | Ok ast -> ast
+    | Error error_msg -> raise (User_error error_msg)
+  in
+  if Ast_editor_state.get_original_mode ast_editor_state then
+    let origin_json =
+      let text = TextDocument.getText document () in
+      match Pp_path.get_kind ~document with
+      | Structure -> make_value (Dumpast.transform text `Impl)
+      | Signature -> make_value (Dumpast.transform text `Intf)
+      | Unknown -> raise (User_error "Unknown file extension")
+    in
+    send_msg "ast" (Jsonoo.t_to_js origin_json) ~webview
+  else
+    let pp_value =
+      try
+        match Pp_path.get_pp_path ~document with
+        | None -> raise (User_error "Project root path wasn't found")
+        | Some path -> (
+          match get_preprocessed_ast path with
+          | Error err_msg -> raise (User_error err_msg)
+          | Ok res ->
+            let pp_json =
+              let pp_code = fetch_pp_code ~document in
+              let lex = Lexing.from_string pp_code in
+              Result.ok_or_failwith
+                (match Ppxlib.Ast_io.get_ast res with
+                | Impl ppml_structure ->
+                  let reparsed_structure = Parse.implementation lex in
+                  Dumpast.reparse ppml_structure reparsed_structure
+                | Intf signature ->
+                  let reparsed_signature = Parse.interface lex in
+                  Dumpast.reparse_signature signature reparsed_signature)
+            in
+            make_value (Ok pp_json))
+      with
+      | Sys_error e -> raise (User_error e)
```

Comment:
Can we localize catching this `Sys_error`? It's not clear who is actually throwing this.

Reply:
The `Sys_error` here occurs when you give inexistent path to `Ppxlib.Ast_io.read_binary`. What do you mean by localizing it, like print the call stack or maybe catch it earlier and add a clarifying message?

Reply:
I pushed a commit that made `Sys_error ` part of the result types. This makes it more obvious to see where this error is coming from.